lib/phase/flatten.js

```diff
@@ -522,6 +537,75 @@ function convertSteps(jobs) {
     return newJobs;
 }
 
+/**
+ * Get full stage name
+ * @param  {String} stageName               Stage name
+ * @param  {String} type                    Type of stage job (setup or teardown)
+ * @return {String}                         Full stage name (e.g. stage@deploy:setup)
+ */
+function getFullStageName({ stageName, type }) {
+    return `${STAGE_PREFIX}${stageName}:${type}`;
+}
+
+/**
+ * Flatten stage setup and teardown into jobs object
+ * @param  {Object} doc               Document that went through structural parsing
+ */
+function flattenStageSetupAndTeardownJobs(doc) {
+    const newJobs = clone(doc.jobs);
+    const stages = clone(doc.stages);
+
+    if (stages) {
+        Object.keys(stages).forEach(stageName => {
+            const { setup, teardown, jobs: stageJobs } = stages[stageName];
+
+            stageJobs.forEach(jobName => {
+                if (newJobs[jobName]) {
+                    newJobs[jobName].stage = { name: stageName };
+                    // Handle startFrom
+                    if (
+                        newJobs[jobName].requires &&
+                        newJobs[jobName].requires.includes(`~${getFullStageName({ stageName, type: 'setup' })}`)
```

Comment:
We should not allow the prefix `~` for `setup` in `requires` of a stage job.

Let say a stage `deploy` has a job with requires like `[~stage@deploy:setup, ~jobOutsideStage]`.
This would mean the stage job can be executed after `jobOutsideStage` is completed without waiting for `~stage@deploy:setup`